src/lib/chawathe98_matcher.rs

```diff
@@ -46,35 +46,2262 @@
+        new_value: usize,
+        new_edge_type: EdgeType,
+    ) {
+        let new_edge = Edge::new(new_from_node, new_to_node, new_value, new_edge_type);
+        self.list_edges.borrow_mut().push(new_edge);
+    }
+    /// Remove Edge
+    pub fn remove(&mut self, index: usize) {
+        self.list_edges.borrow_mut().remove(index);
+    }
+    /// Check contains the from_node
+    pub fn contains_from(&self, from: NodeId<FromNodeId>) -> bool {
+        let list_edges = self.list_edges.borrow();
+        for edge in list_edges.iter() {
+            if edge.from_node == from {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /// Check if the edges contains
+    ///     1) from node id &
+    ///     2) To node id
+    ///     3) Edge Type
+    pub fn contains_edge(
+        &self,
+        from: usize,
+        to: usize,
+        edge_type: EdgeType,
+        list_edges: &mut Vec<Edge>,
+    ) -> bool {
+        for edge in list_edges {
+            if edge.from_node.id() == from && edge.to_node.id() == to && edge.edge_type == edge_type
+                {
+                    return true;
+                }
+        }
+        return false;
+    }
+
+    /// Check if the edge contains the to_node
+    pub fn contains_to(&self, to: NodeId<ToNodeId>) -> bool {
+        let list_edges = self.list_edges.borrow();
+        for edge in list_edges.iter() {
+            if edge.to_node == to {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /// Print for all the pre order traversal tree2: <Vec<NodeId<ToNodeId>>>
+    pub fn print_preorder_traversal(
+        &self,
+        tree1: Vec<NodeId<FromNodeId>>,
+        tree2: Vec<NodeId<ToNodeId>>,
+    ) -> () {
+        println!();
+        println!("Tree 1");
+        print!("|");
+        for node in tree1 {
+            print!("{:?},{:?}|", node.id(), self.from_arena.borrow()[node].label);
+        }
+        println!();
+        println!("Tree 2");
+        print!("|");
+        for node in tree2 {
+            print!("{:?},{:?}|", node.id(), self.to_arena.borrow()[node].label);
+        }
+        println!();
+    }
+    /// Checks if the subtree in node id in From_node is the same same in to_node
+    /// This overall checks the subtrees of two trees and whether they are the same.
+    /// The check is done by checking if the nodes are equal and as well as their descendants.
+    pub fn check_subtree(&self, from_node: NodeId<FromNodeId>, to_node: NodeId<ToNodeId>) -> bool {
+        let mut output = true;
+        let descendants_from_node = from_node
+            .descendants(&self.from_arena.borrow())
+            .collect::<Vec<NodeId<FromNodeId>>>();
+        let descendants_to_node = to_node
+            .descendants(&self.to_arena.borrow())
+            .collect::<Vec<NodeId<ToNodeId>>>();
+        if eq(&from_node, &self.from_arena.borrow(), &to_node, &self.to_arena.borrow()) {
```

Comment:
I'm not sure that I've understood this comparison correctly. Does it compare two `ast::Node` types? If so, I doubt that is what you want to do. The `Node` struct looks like this (I've removed the comments):

```rust
pub struct Node<T: Clone, U: PartialEq + Copy> {
    parent: Option<NodeId<U>>,
    previous_sibling: Option<NodeId<U>>,
    next_sibling: Option<NodeId<U>>,
    first_child: Option<NodeId<U>>,
    last_child: Option<NodeId<U>>,
    index: Option<NodeId<U>>,
    pub label: String,
    pub ty: T,
    pub char_no: Option<usize>,
    pub col_no: Option<usize>,
    pub line_no: Option<usize>,
    pub token_len: Option<usize>,
}
```

So, much of what you are comparing is `NodeId` ids, but subtrees can be "the same" in the two ASTs even if their node ids are different. I think what you are looking for here is something like the `is_isomorphic()` function in the `matchers.rs` file? Either way, this is a function that will need careful testing!